---
title: "Extended Analysis: Merge then Select"
author: "Hubert Rehrauer / Jose F Carre√±o"
output: 
  html_document:
    self_contained: true
    includes:
     in_header: !expr system.file("templates/fgcz_header.html", package="ezRun", lib.loc=.libPaths())
    css: !expr system.file("templates/fgcz.css", package="ezRun", lib.loc=.libPaths())
    code_folding: hide
editor_options: 
  chunk_output_type: inline
---


#  {.tabset}

## Overview

```{r, warning=FALSE, include=FALSE}
### Load necessary libraries

library(Seurat)
library(cowplot)
library(ggplot2)
library(dplyr) 
library(ezRun)
```

The markers for each cluster is in the file [posMarkers.xlsx](posMarkers.xlsx) and the markers differentiating the HC cluster vs the rest of clusters is at, [hcMarkers.xlsx](hcMarkers.xlsx). Finally, the marker analysis for HC cluster vs the rest in the complete merged dataset is at [hcMarkers_allCells.xlsx](hcMarkers_allCells.xlsx)

  *   The interactive explorer for the merged dataset is at <https://fgcz-shiny.uzh.ch/fgcz_iSEE/?data=p29781/additional-analyses/2023-01-27-Extended-MergeThenSelect/scData_combined.rds> and <https://fgcz-shiny.uzh.ch/fgcz_exploreSeurat_app/?data=p29781/additional-analyses/2023-01-27-Extended-MergeThenSelect/scData_combined.rds>.
  *   The interactive explorer for the selected cells after the merge is at: <https://fgcz-shiny.uzh.ch/fgcz_iSEE/?data=p29781/additional-analyses/2023-01-27-Extended-MergeThenSelect/scData_combined_selected.rds> and <https://fgcz-shiny.uzh.ch/fgcz_exploreSeurat_app/?data=p29781/additional-analyses/2023-01-27-Extended-MergeThenSelect/scData_combined_selected.rds>.
  *   The interactive explorer for the reclustered merged dataset (HCs appear): <https://fgcz-shiny.uzh.ch/fgcz_iSEE/?data=p29781/additional-analyses/2023-01-27-Extended-MergeThenSelect/scData_combined_reclustered.rds> and <https://fgcz-shiny.uzh.ch/fgcz_exploreSeurat_app/?data=p29781/additional-analyses/2023-01-27-Extended-MergeThenSelect/scData_combined_reclustered.rds>.

This report is intended to extend the previous analysis "Merge then Select" with the points discussed on 27/01/2023. The entire pipeline followed can be summarized into:

  *   Merge both datasets
  *   Select cells with a given gene expression pattern + OEP cells in the merged dataset
  *   Reclustering of selected cells
  *   Markers Discovery
  *   Subsequent marker analysis
      -   General Hair Cells
      -   Hair Cells
      -   Cochlear Hair Cells
      -   Vestibular Hair Cells
      -   Otic Epithelium
        
The data used for this analysis is coming from experiment 45 & 46.


Important references are:
  *   Dataset integration algorithm: [https://doi.org/10.1016/j.cell.2019.05.031](https://doi.org/10.1016/j.cell.2019.05.031)
  *   Calculate module score reference: [https://doi.org/10.1126%2Fscience.aad0501](https://doi.org/10.1126%2Fscience.aad0501)
  *   Calculate module score tutorial: [Link](https://www.waltermuskovic.com/2021/04/15/seurat-s-addmodulescore-function/#:~:text=Calculate%20the%20average%20expression%20levels,randomly%20selected%20from%20each%20bin.) (I found this particularly helpful to understand the actual implementation of the algorithm in R)
  
Some comments: 
  *   During cell selection, there are cells belonging to the cluster of EP cells that are not discarded. They are in fact, above the 0.3 threshold for the module 2. Nevertheless, these cells are very well identified in the UMAP of the selected cells, and they form a separate cluster (cluster 4).
  *   The low number of clusters identified in after the selection is due to the granularity of the clustering algorithm, which is less than what was set in previous analysis
  *   The marker analysis of the selected cells has been done by taking into account the original labeling and the new labeling of the HC (i.e. the clustering information of the selected cells has not been considered)


### Set important paths

```{r}
dataDirs <- list(
   day60_CD4 = list(
     cellRanger =
"/srv/gstore/projects/p29781/o30306_CellRangerCount_2022-12-20--11-08-02/IEOday60_CD4protocol",
     seuratReport =
"/srv/gstore/projects/p29781/o30306_ScSeurat_2022-12-20--22-03-42/IEOday60_CD4protocol_SCReport",
     labeledObject = NULL
   ), # Not used for now
###############################################################
   day60_ctrl = list(
     cellRanger =
"/srv/gstore/projects/p29781/o30306_CellRangerCount_2022-12-20--11-08-02/IEOday60_controlprotocol",
     seuratReport =
"/srv/gstore/projects/p29781/o30306_ScSeurat_2022-12-20--22-03-42/IEOday60_controlprotocol_SCReport",
     labeledObject =
"/srv/gstore/projects/p29781/additional-analyses/2022-11-28-relabel-day60-o30306/day60_ctrl"
   ), # This is the second run of the ctrl experiment
   day60_otic = list(
     cellRanger =
"/srv/gstore/projects/p29781/o29804_CellRangerCount_2022-11-04--11-20-44/day60oticdiff_Library_418886_1",
     seuratReport =
"/srv/gstore/projects/p29781/o29804_ScSeurat_2022-11-08--10-01-51/day60oticdiff_Library_418886_1_SCReport",
     labeledObject =
"/srv/gstore/projects/p29781/additional-analyses/2022-11-28-relabel-day60"
   ), # This is the dataset contaminated with neurons
###############################################################
   day30 = list(
     cellRanger =
"/srv/gstore/projects/p27889/o29467_CellRangerCount_2022-09-16--09-26-33/day30oticvesicle_Library_412321",
     seuratReport =
"/srv/gstore/projects/p27889/o29467_ScSeurat_2022-09-22--15-04-32/day30oticvesicle_Library_412321_SCReport",
     labeledObject =
"/srv/gstore/projects/p27889/additional-analyses/2022-10-04-relabel-day30oticvesicle"
   ), # Not used for now
   day08 = list(
     cellRanger =
"/srv/gstore/projects/p27889/o28146_CellRangerCount_2022-07-11--09-59-01/placodeday8test1",
     seuratReport =
"/srv/gstore/projects/p27889/o28146_ScSeurat_2022-05-19--11-14-04/placodeday8test1_SCReport",
     labeledObject =
"/srv/gstore/projects/p27889/additional-analyses/2022-08-19-Seurat-relabeled"
   ) # Not used for now
)
#dataDirs

resultDir <- "/home/jcarreno/sta426_project/results/2023-01-27-Extended-MergeThenSelect"
```

### Genes of interest

This list of genes has been obtained from "list for clustering 7.12.2022.xlsx"

```{r}
reference_genes <- list(
  ghc = c(
'ATOH1',
'MYO7A',
'MYO6',
'PCP4',
'ANXA4',
'GFI1',
'ESPN',
'TMPRSS3',
'BDNF',
'CCER2',
'GNG8',
'POU4F3',
'OTOF',
'FSIP1',
'ZBBX',
'SKOR1',
'DNAH5',
'SLC26A5',
'GATA3',
'LMOD3',
'FGF8',
'INSM1',
'DNM3'
), # General Hair Cells: All markers
hc = c(
'ATOH1',
'MYO7A',
'MYO6',
'PCP4',
'ANXA4',
'GFI1',
'ESPN',
'TMPRSS3',
'BDNF',
'CCER2',
'GNG8',
'POU4F3',
'OTOF'), # Hair Cells
vhc = c(
'FSIP1',
'ZBBX',
'SKOR1',
'DNAH5'), #Vestibular Hair Cells
chc = c(
 'SLC26A5',
'GATA3',
'LMOD3',
'FGF8',
'INSM1',
'DNM3',
'TMPRSS3'),  #Cochlear Hair Cells
oep = c(
'EPCAM',
'CDH1',
'SOX2',
'SIX1',
'OC90',
'SOX10',
'FBXO2',
'LMX1A',
'PAX2',
'PAX8',
'DLX5',
'GBX2',
'JAG1',
'TBX2',
'COL9A2',
'OTOA',
'MYO5C',
'OTOL1',
'USH1C',
'PCDH9')
)
```

### Define colors

```{r}
seuratPal <- scales::hue_pal()(14)
#scales::show_col(seuratPal)
ctColors <- c(
  "EP"=seuratPal[2],
  "NE"=seuratPal[9],
  "MES"=seuratPal[12],
  "NEURONS"="steelblue",
  "OEP"=seuratPal[4],
  "MUSCLE"="gray40",
  "NE-like"=seuratPal[8],
  "ChP"=seuratPal[11],
  "MEL"="gray60",
  "Chondro"="gray80",
  "HC" = "red")
ezShowCol <- function (colours, colorLabels=names(colours), borders = NULL, cex_label = 1, 
          ncol = NULL) 
{
  n <- length(colours)
  ncol <- ncol %||% ceiling(sqrt(length(colours)))
  nrow <- ceiling(n/ncol)
  colours <- c(colours, rep(NA, nrow * ncol - length(colours)))
  colours <- matrix(colours, ncol = ncol, byrow = FALSE)
  old <- par(pty = "s", mar = c(0, 0, 0, 0))
  on.exit(par(old))
  size <- max(dim(colours))
  plot(c(0, size), c(0, -size), type = "n", xlab = "", ylab = "", 
       axes = FALSE)
  rect(col(colours) - 1, -row(colours) + 1, col(colours), 
       -row(colours), col = colours, border = borders)
  if (!is.null(colorLabels)) {
    hcl <- farver::decode_colour(colours, "rgb", "hcl")
    label_col <- ifelse(hcl[, "l"] > 50, "black", "white")
    text(col(colours)[1:length(colorLabels)] - 0.5, -row(colours)[1:length(colorLabels)] + 0.5, colorLabels, 
         cex = cex_label, col = label_col)
  }
}

ezShowCol(ctColors, colorLabels=paste0(names(ctColors), "\n", ctColors))
```

Load the data from the server and identify the source of each dataset

```{r}
day60_otic <- readRDS(file = paste(dataDirs$day60_otic$labeledObject, "/scData.rds",sep = ""))
day60_ctrl <- readRDS(file = paste(dataDirs$day60_ctrl$labeledObject, "/scData.rds",sep = ""))

day60_otic$experiment <- "day60_otic"
day60_ctrl$experiment <- "day60_ctrl"
```


## Merge the datasets

```{r}
day60_otic <- FindVariableFeatures(day60_otic, nfeatures = 2000)
day60_ctrl <- FindVariableFeatures(day60_ctrl, nfeatures = 2000)
```


```{r}
anchors <- FindIntegrationAnchors(object.list = list(day60_otic, day60_ctrl), dims = 1:20, verbose = FALSE)
```

Integrate the selected cells from both datasets:

```{r}
combined <- IntegrateData(anchorset = anchors, dims = 1:20, verbose = FALSE)
```

```{r}
#ctColors <- ctColors[levels(Idents(combined))]
```


From our set of genes stablished in the Overview, check how many of these ARE NOT present in the "integrated" assay:

```{r}
sapply(reference_genes, function(rg){
  rg %>% setdiff(rownames(combined)) %>% length()
})
```

From our set of genes stablished in the Overview, check how many of these ARE present in the "integrated" assay:

```{r}
sapply(reference_genes, function(rg){
  rg %>% intersect(rownames(combined)) %>% length()
})
```


### Reclustering

```{r}
combined <- ScaleData(combined, verbose = FALSE)
combined <- FindVariableFeatures(combined, nfeatures = 2000)
combined <- RunPCA(combined, npcs = 30, verbose = FALSE)
ElbowPlot(combined)
```

```{r}
combined <- RunTSNE(combined, reduction = "pca", dims = 1:5)
combined <- RunUMAP(combined, reduction = "pca", dims = 1:20, verbose = FALSE)
```


```{r}
DimPlot(combined, label = TRUE, cols=ctColors)
```

```{r}
DimPlot(combined, group.by = "Condition")
```

```{r, fig.width=12, fig.height=5}
DimPlot(combined, split.by = "Condition", cols=ctColors)
```

## Select cells

```{r}
DefaultAssay(combined) <- "RNA"
combined <- ScaleData(combined, verbose = FALSE, features=rownames(combined))
```

We compute the actual score: 

```{r}
combined <- AddModuleScore(combined, features=reference_genes, name="module")
```

### Plot scores

```{r, fig.width=20, fig.height=15}
featNames <- paste0("module", 1:length(reference_genes))
p1 <- DimPlot(combined, group.by = "clusterLabels", cols=ctColors)
p2 <- FeaturePlot(combined, featNames)
p3 <- VlnPlot(combined, featNames, group.by = "clusterLabels", ncol = 5, cols=ctColors)

upper_row <- plot_grid(p1, p2, ncol = 2)

plot_grid(upper_row, p3, ncol = 1)
```

```{r}
for (i in 1:length(reference_genes)){
  if(i==2){
    print(VlnPlot(combined, featNames[i], pt.size=1, group.by = "clusterLabels", cols=ctColors) + geom_hline(yintercept=0.30) + ggtitle(names(reference_genes)[i]))
  }
  else if (i==5) {
    print(VlnPlot(combined, featNames[i], pt.size=1, group.by = "clusterLabels", cols=ctColors) + geom_hline(yintercept=0.1) + ggtitle(names(reference_genes)[i]))
  } 
  else {
     print(VlnPlot(combined, featNames[i], pt.size=1, group.by = "clusterLabels", cols=ctColors) + ggtitle(names(reference_genes)[i]))
  }
}
```

Use only the cells above the threshold in module 2 and the otic epithelium cells above threshold in module 5:


```{r}
useCell <- (combined$clusterLabels == "OEP" & combined$module5>0.1) | combined$module2 > 0.30 
combined_sel <- combined[ , useCell]
```

### Reclustering of the selected cells

```{r}
combined_sel <- FindVariableFeatures(combined_sel, nfeatures = 2000)
combined_sel <- RunPCA(combined_sel, npcs = 10, verbose = FALSE)
ElbowPlot(combined_sel)
```

```{r}
combined_sel <- RunUMAP(combined_sel, reduction = "pca", dims = 1:5)
```

```{r}
combined_sel <- RunTSNE(combined_sel, reduction = "pca", dims = 1:5)
combined_sel <- FindNeighbors(combined_sel, reduction = "pca", dims = 1:5, k.param = 7)
```

```{r}
combined_sel <- FindClusters(combined_sel, resolution = 0.3)
```

```{r}
DimPlot(combined_sel, label = TRUE)
```

```{r}
DimPlot(combined_sel, group.by = "Condition")
```

```{r}
DimPlot(combined_sel, group.by = "clusterLabels", cols=ctColors)
```

## Module Score

Compute again the score for each set of genes established in the Overview section:

```{r, fig.width=10, fig.height=10}
featNames <- paste0("module", 1:length(reference_genes))
FeaturePlot(combined_sel, featNames, order= TRUE)
```

Plot the set of genes stablished in the Overview and that are present in the integrated object:

```{r, fig.width=10, fig.height=10}
rgSet <- lapply(reference_genes, intersect, rownames(combined_sel))
DoHeatmap(combined_sel, unlist(rgSet))
```


Number of cells per cluster:

```{r}
cellCountperCluster <- data.frame(id = Idents(combined_sel))
barplot = ggplot(data=cellCountperCluster, aes(x=id)) + 
  geom_bar() +
  geom_text(stat='count', aes(label=..count..), vjust=-1) +
  theme_minimal()
barplot + labs(x="Cluster", y = "Cell count")
```


Cells in cluster 5 are likely to be HC:

```{r}
Idents(combined_sel) <- combined_sel$seurat_clusters %>% recode("5"="HC")
```


HIGHLIGHT VERSION 1: ALL GREEN VS HC IN RED

```{r}
cells.use <- WhichCells(combined_sel, idents = c('HC'))
DimPlot(combined_sel, reduction = "umap", cells.highlight = cells.use, sizes.highlight = 0.5, cols=seuratPal[4]) + NoLegend()
```


HIGHLIGHT VERSION 2: ALL CELLS THEIR RESPECTIVE COLORS & HC RED

```{r}
Idents(combined_sel) <- combined_sel$clusterLabels
combined_sel <- SetIdent(combined_sel, cells = cells.use, value = 'HC')
```


```{r}
DimPlot(combined_sel, cols=ctColors)
```

HIGHLIGHT VERSION 3: ALL GRAY VS HC IN RED

```{r}
DimPlot(combined_sel, reduction = "umap", cells.highlight = cells.use, sizes.highlight = 0.5) + NoLegend()
```
## Mapping HC to original Merged Dataset

```{r, fig.width=12, fig.height=4}
p1 <- DimPlot(combined, cols=ctColors)
combined_reclustered <- SetIdent(combined, cells = cells.use, value = 'HC')
p2 <- DimPlot(combined_reclustered, cols=ctColors)

plot_grid(p1, p2, ncol=2)
```


## Markers Discovery {.tabset}

### All vs all

```{r}
markers <- FindAllMarkers(object=combined_sel, test.use = "wilcox", only.pos=TRUE, assay="RNA")
## Significant markers
markers <- markers[ ,c("gene","cluster","pct.1", "pct.2", "avg_log2FC","p_val_adj")]
#cm <- cm[cm$p_val_adj < 0.05, ]
markers$cluster <- as.factor(markers$cluster)
markers$diff_pct = abs(markers$pct.1-markers$pct.2)
markers <- markers[order(markers$diff_pct, decreasing = TRUE),] ## why would we round here?? %>% mutate_if(is.numeric, round, digits=3)
writexl::write_xlsx(markers, path=file.path(resultDir, "posMarkers.xlsx"))
```

```{r heatmap, fig.width=12, fig.height=12}
top5 <- markers %>% group_by(cluster) 
top5 <- slice_max(top5, n = 5, order_by = avg_log2FC)
genesToPlot <- unique(as.character(top5$gene))
genesToPlot <- intersect(genesToPlot, rownames(combined_sel))

DoHeatmap(combined_sel, features=unique(genesToPlot))
DotPlot(combined_sel, features=genesToPlot) + coord_flip()
```

### HC vs Rest

```{r}

markers <- FindMarkers(object=combined_sel, ident.1 = "HC", test.use = "wilcox", assay="RNA")
## Significant markers
markers <- ezFrame(gene=rownames(markers), markers)
markers <- markers[ ,c("gene","pct.1", "pct.2", "avg_log2FC","p_val_adj")]
#cm <- cm[cm$p_val_adj < 0.05, ]
markers$diff_pct = abs(markers$pct.1-markers$pct.2)
markers <- markers[order(markers$diff_pct, decreasing = TRUE),] ## why would we round here?? %>% mutate_if(is.numeric, round, digits=3)
writexl::write_xlsx(markers, path=file.path(resultDir, "hcMarkers.xlsx"))
```

```{r , fig.width=12, fig.height=12}

genesToPlot <- c(rownames(markers[order(-markers$avg_log2FC), ]) %>% head(30),
                 rownames(markers[order(markers$avg_log2FC), ]) %>% head(30)) %>% unique()


DoHeatmap(combined_sel, features=unique(genesToPlot))
DotPlot(combined_sel, features=genesToPlot) + coord_flip()
```

### HC vs Rest (complete merged dataset)

```{r}

markers <- FindMarkers(object=combined_reclustered, ident.1 = "HC", test.use = "wilcox", assay="RNA")
## Significant markers
markers <- ezFrame(gene=rownames(markers), markers)
markers <- markers[ ,c("gene","pct.1", "pct.2", "avg_log2FC","p_val_adj")]
#cm <- cm[cm$p_val_adj < 0.05, ]
markers$diff_pct = abs(markers$pct.1-markers$pct.2)
markers <- markers[order(markers$diff_pct, decreasing = TRUE),] ## why would we round here?? %>% mutate_if(is.numeric, round, digits=3)
writexl::write_xlsx(markers, path=file.path(resultDir, "hcMarkers_allCells.xlsx"))
```

Differentially Expressed Genes for the complete merged dataset. Please, note that the second histogram is a subsampled version of the first one to improve visualization.

```{r , fig.width=15, fig.height=12}

genesToPlot <- c(rownames(markers[order(-markers$avg_log2FC), ]) %>% head(30),
                 rownames(markers[order(markers$avg_log2FC), ]) %>% head(30)) %>% unique()

maxcells  <- 3*min(table(Idents(combined_reclustered)))

DoHeatmap(combined_reclustered, features=unique(genesToPlot))
DoHeatmap(subset(combined_reclustered, downsample = maxcells), features=unique(genesToPlot))
DotPlot(combined_reclustered, features=genesToPlot) + coord_flip()
```

## Markers Analysis {.tabset}

### General Hair Cells Markers {.tabset}

#### Heatmap

```{r, fig.height=10, fig.width=8}
DoHeatmap(combined_sel, reference_genes$ghc)
```


#### UMAPs

```{r}
for (marker in reference_genes$ghc){
  tryCatch({print(FeaturePlot(combined_sel, reduction = "umap", features = marker, order= TRUE))},
           error=function(e){print("ERROR")},
           warning=function(w){print(paste(marker, "not present"))})
}
```


#### Dotplot

```{r}
DotPlot(combined_sel, features=reference_genes$ghc) + coord_flip() + theme(axis.text.x = element_text(size = 8))  
```


#### Violin Plots


```{r}
for (marker in reference_genes$ghc){
  print(VlnPlot(combined_sel, marker, cols=ctColors))
  #tryCatch({print(VlnPlot(combined_sel, marker, cols=ctColors))},
  #        error=function(e){print("ERROR")},
  #        warning=function(w){print(VlnPlot(combined_sel, marker, cols=ctColors))})
}
```



### Hair Cells Markers {.tabset}

#### Heatmap

```{r, fig.height=8, fig.width=8}
DoHeatmap(combined_sel, reference_genes$hc)
```


#### UMAPs

```{r, fig.width=5, fig.height=4}
for (marker in reference_genes$hc){
  tryCatch({print(FeaturePlot(combined_sel, reduction = "umap", features = marker, order= TRUE))},
           error=function(e){print("ERROR")},
           warning=function(w){print(paste(marker, "not present"))})
}
```


#### Dotplot

```{r}
DotPlot(combined_sel, features=reference_genes$hc) + coord_flip() + theme(axis.text.x = element_text(size = 8))  
```


#### Violin Plots


```{r, fig.width=5, fig.height=4}
for (marker in reference_genes$hc){
  print(VlnPlot(combined_sel, marker, cols=ctColors))
  #tryCatch({print(VlnPlot(combined_sel, marker, cols=ctColors))},
  #         error=function(e){print("ERROR")},
  #         warning=function(w){print(paste(marker, "not present"))})
}
```



### Cochlear Hair Cells Markers {.tabset}

#### Heatmap

```{r, fig.height=8, fig.width=8}
DoHeatmap(combined_sel, reference_genes$chc)
```


#### UMAPs

```{r, fig.width=5, fig.height=4}
for (marker in reference_genes$chc){
  tryCatch({print(FeaturePlot(combined_sel, reduction = "umap", features = marker, order= TRUE))},
           error=function(e){print("ERROR")},
           warning=function(w){print(paste(marker, "not present"))})
}
```


#### Dotplot

```{r}
DotPlot(combined_sel, features=reference_genes$chc) + coord_flip() + theme(axis.text.x = element_text(size = 8))  
```


#### Violin Plots


```{r, fig.width=5, fig.height=4}
for (marker in reference_genes$chc){
  print(VlnPlot(combined_sel, marker, cols=ctColors))
  #tryCatch({print(VlnPlot(combined_sel, marker, cols=ctColors))},
  #         error=function(e){print("ERROR")},
  #         warning=function(w){print(paste(marker, "not present"))})
}
```


### Otic Epithelium Markers {.tabset}

#### Heatmap

```{r, fig.height=8, fig.width=8}
DoHeatmap(combined_sel, reference_genes$oep)
```


#### UMAPs

```{r, fig.width=5, fig.height=4}
for (marker in reference_genes$oep){
  tryCatch({print(FeaturePlot(combined_sel, reduction = "umap", features = marker, order= TRUE))},
           error=function(e){print("ERROR")},
           warning=function(w){print(paste(marker, "not present"))})
}
```


#### Dotplot

```{r}
DotPlot(combined_sel, features=reference_genes$oep) + coord_flip() + theme(axis.text.x = element_text(size = 8))  
```


#### Violin Plots


```{r, fig.width=5, fig.height=4}
for (marker in reference_genes$oep){
  print(VlnPlot(combined_sel, marker, cols=ctColors))

  #tryCatch({print(VlnPlot(combined_sel, marker, cols=ctColors))},
  #         error=function(e){print("ERROR")},
  #         warning=function(w){print(paste(marker, "not present"))})
}
```



## SessionInfo

```{r, echo=FALSE}
saveRDS(combined, file = file.path(resultDir, "scData_combined.rds"))
saveRDS(combined_sel, file = file.path(resultDir, "scData_combined_selected.rds"))
saveRDS(combined_reclustered, file = file.path(resultDir, "scData_combined_reclustered.rds"))

sessionInfo()
```
